### ISSUE TITLE
ExcludeFromAutoload seems to have no effect

### ISSUE THREAD

**Issue by dimpase** [gap-system/gap#5230]
### Observed behaviour
```
% cat ~/.gap/gaprc 
SetUserPreference("ExcludeFromAutoload", ["atlasrep", "io"]);
```

yet, `io` and `atlasrep` are loaded at startup

### Expected behaviour

no `io` and `atlasrep` loaded at startup

### Copy and paste GAP banner (to tell us about your setup)
```
% ./gap               
 ┌───────┐   GAP 4.13dev-169-g28d5faa built on 2022-11-30 18:44:58+0000
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin22-default64-kv8
 Configuration:  gmp 6.2.1, GASMAN, readline
 Loading the library and packages ...
 Packages:   AClib 1.3.2, Alnuth 3.2.1, AtlasRep 2.1.6, AutPGrp 1.11, 
             CRISP 1.4.5, Cryst 4.1.25, CrystCat 1.1.10, CTblLib 1.3.4, 
             FactInt 1.6.3, FGA 1.4.0, GAPDoc 1.6.6, IRREDSOL 1.4.4, 
             LAGUNA 3.9.5, Polenta 1.3.10, Polycyclic 2.16, 
             PrimGrp 3.4.2, RadiRoot 2.9, ResClasses 4.7.3, 
             SmallGrp 1.5.1, Sophus 1.27, SpinSym 1.5.2, TomLib 1.2.9, 
             TransGrp 3.6.3, utils 0.78
 Try '??help' for help. See also '?copyright', '?cite' and '?authors'
gap> UserPreference( "ExcludeFromAutoload" );
[ "atlasrep", "io" ]
```

I think I checked that `io` is not listed as `Needed...` by any of these packages listed above. What am I doing wrong?

**Comment by dimpase** [comment]
The reason I ask is because libgap seems to be broken on macOS 13.0.1 if `io` is built, as can be seen by running
`tst/testlibgap/basic`:
```
% ./tst/testlibgap/basic -l .
# Initializing GAP...
 ┌───────┐   GAP 4.13dev-169-g28d5faa built on 2022-11-30 18:44:58+0000
 │  GAP  │   https://www.gap-system.org
 └───────┘   Architecture: x86_64-apple-darwin22-default64-kv8
 Configuration:  gmp 6.2.1, GASMAN, readline
 Loading the library and packages ...
Error, LOAD_DYN: failed to load kernel module ./pkg/io//bin/x86_64-apple-darwin2\
2-default64-kv8/io.so, dlopen(./pkg/io//bin/x86_64-apple-darwin22-default\
64-kv8/io.so, 0x0005): Symbol not found: _AssGVar
  Referenced from: <691ED98B-03DB-3A1C-98A6-D5307CB4FCF6> /Users/dima/sof\
tware/gap/pkg/io/bin/x86_64-apple-darwin22-default64-kv8/io.so
  Expected in:     <6EA6D1DA-9BAB-322E-AEE4-887EFCA0E1EF> /Users/dima/sof\
tware/gap/tst/testlibgap/.libs/basic in
  LOAD_DYN( filename ) at ./lib/files.gd:595 called from 
<function "LoadDynamicModule">( <arguments> )
 called from read-eval loop at ./pkg/io//init.g:23
Error, was not in any namespace at ./lib/variable.g:269 called from
LEAVE_NAMESPACE(  ); at ./lib/package.gi:1327 called from
ReadPackage( pkgname, "init.g" ); at ./lib/package.gi:1688 called from
LoadPackage( pair[1], pair[2], false ) at ./lib/package.gi:1933 called from
func(  ); at ./lib/system.g:228 called from
<function "CallAndInstallPostRestore">( <arguments> )
 called from read-eval loop at ./lib/init.g:700
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk> 
```

**Comment by dimpase** [comment]
noted while trying to upgrade Sage's GAP on
https://trac.sagemath.org/ticket/34391

**Comment by dimpase** [comment]
also, it can be triggered non-interactively. With the following change
```c
--- a/tst/testlibgap/basic.c
+++ b/tst/testlibgap/basic.c
@@ -6,6 +6,7 @@ int main(int argc, char ** argv)
 {
     printf("# Initializing GAP...\n");
     GAP_Initialize(argc, argv, 0, 0, 1);
+    test_eval("LoadPackage(\"io\");");
     test_eval("1+2+3;");
     test_eval("g:=FreeGroup(2);");
     test_eval("a:=g.1;");
```

by running `make testlibgap` on macOS 13 one gets the same  error as in https://github.com/gap-system/gap/issues/5230#issuecomment-1332721919.

I'm not sure how to make this into a proper test, as this needs `io` package compiled.

**Comment by ThomasBreuer** [comment]
The user's `gaprc` file is read *after* autoloading packages.
The user preference in question can be set in `gap.ini`,
see [the documentation about these files](https://docs.gap-system.org/doc/ref/chap3_mj.html#X7FD66F977A3B02DF).

**Comment by dimpase** [comment]
I'll split this issue into two, sorry for putting two different issues into one.

**Comment by dimpase** [comment]
> The user's `gaprc` file is read _after_ autoloading packages.
> The user preference in question can be set in `gap.ini`,
> see [the documentation about these files](https://docs.gap-system.org/doc/ref/chap3_mj.html#X7FD66F977A3B02DF).

I was misled by an advice to move `~/.gaprc` (yes, I am a long, long term GAP user :-)) to `~/.gap/gaprc` I read somewhere.
The docs are not totally clear on meaningless of putting `SetUserPreference("ExcludeFromAutoload", [...]);` into `.gap/gaprc`. Anyway, having it in the latter should not even change `ExcludeFromAutoload` value, as ability to change it where it makes no sense to do so is misleading. Better even make it an error.

Last but not the least, is there any explicit place in the GAP where the list of autoloaded packages is set up? I could not find it. Should there be a user-controlled way to form such a list?

**Comment by dimpase** [comment]
Let's discuss libgap on #5232.

**Comment by ThomasBreuer** [comment]
> I was misled by an advice to move ~/.gaprc (yes, I am a long, long term GAP user :-)) to ~/.gap/gaprc I read somewhere.
The docs are not totally clear on meaningless of putting SetUserPreference("ExcludeFromAutoload", [...]); into .gap/gaprc. Anyway, having it in the latter should not even change ExcludeFromAutoload value, as ability to change it where it makes no sense to do so is misleading. Better even make it an error.

[The documentation about gap.ini and gaprc](https://docs.gap-system.org/doc/ref/chap3_mj.html#X7FD66F977A3B02DF) says clearly that `gap.ini` is read before packages get autoloaded and `gaprc` is read afterwards. Thus it is clear that the list of autoloaded packages cannot be controlled by the contents of `gaprc`.

User preferences can be set at any time in a GAP session, and this can make sense even for the preference `ExcludeFromAutoload`, for example when one calls `WriteGapIniFile` afterwards, such that the new values become relevant for the next GAP session.

**Comment by dimpase** [comment]
> User preferences can be set at any time in a GAP session, and this can make sense even for the preference `ExcludeFromAutoload`, for example when one calls `WriteGapIniFile` afterwards, such that the new values become relevant for the next GAP session.

Isn't it possible to block/unblopck changing `ExcludeFromAutoload`? If so, blocking can be done after `gap.ini` is read. If one wants to change it, unblock it.

**Comment by ThomasBreuer** [comment]
> Last but not the least, is there any explicit place in the GAP where the list of autoloaded packages is set up? I could not find it. Should there be a user-controlled way to form such a list?

This list can be obtained by calling `UserPreference("PackagesToLoad")`, see [the manual section about "Automatic loading of GAP packages"](https://docs.gap-system.org/doc/ref/chap76_mj.html#X7E6767B485F23BFC).

**Comment by dimpase** [comment]
> `gap.ini` is read before packages get autoloaded

it's not clear which of many possible `gap.ini` is actually read. Can one even have `gap.ini` in `~/.gap/` read? From `The root directories are specified via one or several of the -l paths command line options` in sect. 9.2 one might be lead to think that `~/.gap/` must be specified there.

e.g. on macOS with GAP started as `./gap` from the source directory, I see
```
gap> GAPInfo.RootPaths;
[ "/Users/dima/Library/Preferences/GAP/", "/Users/dima/.gap/", 
  "/Users/dima/software/gap/" ]
```
the 1st entry refers to a non-existing directory. How does it come there?

**Comment by ThomasBreuer** [comment]
@dimpase 

> it's not clear which of many possible gap.ini is actually read.

[The documentation](https://docs.gap-system.org/doc/ref/chap3_mj.html#X7FD66F977A3B02DF) says that the first `gap.ini` that is found in the root directories is read.

> From "The root directories are specified via one or several of the -l paths command line options" in sect. 9.2 one might be lead to think that ~/.gap/ must be specified there.

No.
The next sentence in the paragraph in question is as follows.

> Furthermore, by default GAP automatically prepends a user specific GAP root directory to the list; this can be avoided by  calling GAP with the -r option. The name of this user specific directory depends  on  your  operating system, it can be found in GAPInfo.UserGapRoot.

Concerning the entries in your `GAPInfo.RootPaths`, you can check which entries correspond to `GAPInfo.UserGapRoot` and to `GAPInfo.CommandLineOptions.l`. The second entry in your list looks like `GAPInfo.UserGapRoot`; if this is the case then I do not understand why and how the first entry got prepended.
(Of course one can modify `GAPInfo.RootPaths`, for example with `ExtendRootDirectories`, but this function just appends paths.)

**Comment by fingolfin** [comment]
I don't see a bug here, and the documentation seems clear? E.g. looking at the documentation for `SetUserPreference`, we see: 

> The  easiest  way  to  make  use  of  user preferences is probably to use the function `WriteGapIniFile`, usually
  without   argument.   This  function  creates  a  file  `gap.ini`  in  your  user  specific  GAP  root  directory
  (`GAPInfo.UserGapRoot`).

So, just do

```
SetUserPreference("ExcludeFromAutoload", [...]);
WriteGapIniFile();
```

**Comment by dimpase** [comment]
there is no `ExcludeFromAutoload` in https://docs.gap-system.org/doc/ref/chapInd_mj.html 

there ought to be a section on autoloading and how to manage it in `76 Using and Developing GAP Packages`

**Comment by dimpase** [comment]
> there is no `ExcludeFromAutoload` in https://docs.gap-system.org/doc/ref/chapInd_mj.html
> 
> there ought to be a section on autoloading and how to manage it in `76 Using and Developing GAP Packages`

that's what causes one to search for solutions on the net, bringing up various not up to date tips, instead of a correct way.

**Comment by ThomasBreuer** [comment]
@dimpase There is [a section on autoloading in chapter 76](https://docs.gap-system.org/doc/ref/chap76_mj.html#X7E6767B485F23BFC).

Currently there are indeed no index entries about supported user preferences; it would be good to add such entries.

**Comment by fingolfin** [comment]
I agree that `ExcludeFromAutoload` (and other user preferences) should be findable via `??` and the manual index, and indeed should be listed in the manual. @ThomasBreuer opened a separate issue for that, see #5233.

In addition, there are a few hits for `??automatic load`; perhaps in those places we should add `<Index>autoload</Index>`, too?

**Comment by fingolfin** [comment]
@dimpase so, did you manage to get `ExcludeFromAutoload` working now on the above? Then is there anything left to be done here, given that the suggested documentation improvements are being tracked separately in #5233?

**Comment by dimpase** [comment]
yes, `% mv ~/.gap/gaprc ~/.gap/gap.ini` did the trick, thanks.
